File: src/main/java/edu/cornell/cs/apl/viaduct/errors/CompilationError.java
package edu.cornell.cs.apl.viaduct.errors;

import edu.cornell.cs.apl.viaduct.util.PrintUtil;
import java.io.PrintStream;
import org.apache.commons.lang3.StringUtils;
import org.fusesource.jansi.Ansi;
import org.fusesource.jansi.Ansi.Color;
import org.fusesource.jansi.AnsiPrintStream;

/**
 * Superclass of all errors caused by bad user input to the compiler.
 *
 * <p>Errors caused by bugs in the compiler do not belong here. They should instead raise a
 * (subclass of) {@link RuntimeException}.
 */
public abstract class CompilationError extends Error {
  /** General description (i.e. title) of the error. */
  protected abstract String getCategory();

  /** Name of the file or description of the source that caused the error. */
  protected abstract String getSource();

  /**
   * Print colored error description to the given stream.
   *
   * <p>The output stream should be ready to handle ANSI color codes. If the output stream does not
   * support color codes, they can be stripped by wrapping the output stream in {@link
   * AnsiPrintStream}.
   */
  public void print(PrintStream output) {
    // Print title line by default
    final String title = "-- " + getCategory().toUpperCase() + " - ";
    final int paddingLength = PrintUtil.LINE_WIDTH - getSource().length() - title.length();
    final String padding = StringUtils.repeat('-', paddingLength);

    output.println(Ansi.ansi().fg(Color.CYAN).a(title).a(padding).a(getSource()).reset());
    output.println();
  }

  @Override
  public final String toString() {
    return PrintUtil.printToString(this::print);
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/errors/CompilationError.kt
package edu.cornell.cs.apl.viaduct.errors

import edu.cornell.cs.apl.prettyprinting.AnsiBaseColor
import edu.cornell.cs.apl.prettyprinting.AnsiColor
import edu.cornell.cs.apl.prettyprinting.DEFAULT_LINE_WIDTH
import edu.cornell.cs.apl.prettyprinting.Document
import edu.cornell.cs.apl.prettyprinting.NormalColor
import edu.cornell.cs.apl.prettyprinting.PrettyPrintable
import edu.cornell.cs.apl.prettyprinting.Style
import edu.cornell.cs.apl.prettyprinting.div
import edu.cornell.cs.apl.prettyprinting.nested
import edu.cornell.cs.apl.prettyprinting.plus
import edu.cornell.cs.apl.prettyprinting.styled
import edu.cornell.cs.apl.viaduct.syntax.SourceLocation

/**
 * Superclass of all errors caused by bad user input to the compiler.
 *
 * Errors caused by bugs in the compiler do not belong here.
 * They should instead raise a (subclass of) [RuntimeException].
 */
abstract class CompilationError : Error(), PrettyPrintable {
    /** General description (i.e. title) of the error. */
    protected abstract val category: String

    /** Name of the file or description of the source that contains the error. */
    protected abstract val source: String

    /** Detailed description of the error. */
    protected abstract val description: Document

    /** A hint that can help resolve the error. */
    protected open val hint: Document?
        get() = null

    private object HeaderStyle : Style {
        override val foregroundColor: AnsiColor
            get() = NormalColor(AnsiBaseColor.CYAN)
    }

    private object SourceHighlightingStyle : Style {
        override val foregroundColor: AnsiColor
            get() = NormalColor(AnsiBaseColor.RED)
    }

    /**
     * Displays this message followed by the portion of the source code indicated
     * by [sourceLocation].
     */
    protected fun Document.withSource(sourceLocation: SourceLocation): Document =
        this / (Document.lineBreak + sourceLocation.showInSource(SourceHighlightingStyle))

    /** Displays this message followed by [body] with [body] on its own line. */
    protected fun Document.withData(body: PrettyPrintable): Document =
        this / (Document.lineBreak + body).nested() + Document.lineBreak + Document.lineBreak

    final override val asDocument: Document
        get() {
            val title = "-- ${category.toUpperCase()} -"
            val source = " " + this.source
            val paddingLength = (DEFAULT_LINE_WIDTH - title.length - source.length).coerceAtLeast(0)
            val padding = "-".repeat(paddingLength)
            val header = (Document(title) + padding + source).styled(HeaderStyle)

            val hint = hint?.plus(Document.lineBreak + Document.lineBreak) ?: Document()

            return header + Document.lineBreak + Document.lineBreak + description + hint
        }

    final override fun toString(): String =
        this.asDocument.print()
}
